### PR TITLE
7903239: ofAddress factory of function pointer type is wrong for struct returns

### PR DESCRIPTION
**Pull request by lost-illusi0n**
When generating the lambda inside the `ofAddress` factory, it is missing the `SegmentAllocator` needed when the function returns a struct. This is [implemented for regular functions](https://github.com/openjdk/jextract/blob/0582eaf1b4cdba95f0ee8c2480767433bb647d0d/src/main/java/org/openjdk/jextract/impl/HeaderFileBuilder.java#L122-L127), but not for function pointers.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (no review required)

### Issue
 * [CODETOOLS-7903239](https://bugs.openjdk.org/browse/CODETOOLS-7903239): ofAddress factory of function pointer type is wrong for struct returns (**Bug** - P3) ⚠️ Issue is already resolved. Consider making this a "backport pull request" by setting the PR title to `Backport <hash>` with the hash of the original commit. See [Backports](https://wiki.openjdk.org/display/SKARA/Backports).

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer) 🔄 Re-review required (review was made when pull request targeted the [master](https://git.openjdk.org/jextract/tree/master) branch)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jextract.git pull/58/head:pull/58` \
`$ git checkout pull/58`

Update a local copy of the PR: \
`$ git checkout pull/58` \
`$ git pull https://git.openjdk.org/jextract.git pull/58/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 58`

View PR using the GUI difftool: \
`$ git pr show -t 58`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jextract/pull/58.diff">https://git.openjdk.org/jextract/pull/58.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jextract/pull/58#issuecomment-1200137588)

### CODE REVIEW

File: test/testng/org/openjdk/jextract/test/toolprovider/Test7903239.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import testlib.TestUtils;
+import org.testng.annotations.Test;
+import testlib.JextractToolRunner;
+
+import java.nio.file.Path;
+
+public class Test7903239 extends JextractToolRunner {
+    @Test
+    public void testFunctionPointerReturningStruct() {
```
Comment (mcimadamore) [inline test/testng/org/openjdk/jextract/test/toolprovider/Test7903239.java:34]: While this test in itself is good, I feel it's perhaps the wrong kind of test, in the sense that it only checks that jextract completes normally. Maybe a test that uses the jextract jtreg extension would be better? For inspiration look here:

https://github.com/openjdk/jextract/blob/master/test/jtreg/generator/funcPointerInvokers/TestFuncPointerInvokers.java

You can basically run jextract as part of a jtreg test action, as here:

https://github.com/openjdk/jextract/blob/master/test/jtreg/generator/funcPointerInvokers/TestFuncPointerInvokers.java#L39

Then you can actually compile (against the jextracted bits) and even run:

https://github.com/openjdk/jextract/blob/master/test/jtreg/generator/funcPointerInvokers/TestFuncPointerInvokers.java#L40

So you can check that the generated code actually behaves as expected at runtime.

File: test/jtreg/generator/test7903239/libTest7903239.c [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (mcimadamore) [inline test/jtreg/generator/test7903239/libTest7903239.c:2]: If the test is new, should it have 2022 in the copyright year? (also in c file)

File: test/jtreg/generator/test7903239/Test7903239.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * @run main/othervm JtregJextractSources -l Test7903239 -t test.jextract.test7903239 test7903239.h
+ * @run testng/othervm --enable-native-access=ALL-UNNAMED Test7903239
+ */
+public class Test7903239 {
+    @Test
+    public void testOfAddress() {
+        try (MemorySession session = MemorySession.openConfined()) {
+            var foo = test7903239_h.foo$SEGMENT();
+
+            var barA = Foo.a(foo, session).apply();
```
Comment (mcimadamore) [inline test/jtreg/generator/test7903239/Test7903239.java:51]: I think the limitation of the strategy are clearer in this example :-)
The allocator is decided once and for all when the functional interface instance is obtained. Any other call to `apply` will use the same segment allocator specified on creation. To solve this, I believe the only way is to use the default method setup I described earlier. But the patch is good enough as is for now.

Comment (lost-illusi0n) [reply test/jtreg/generator/test7903239/Test7903239.java:51]: Yes, the tradeoff is quite apparent, but I suppose reasonable to an extent. Your solution is probably the best road to go down when playing around with it though :+1:.

File: test/jtreg/generator/test7903239/libTest7903239.c [hunk 4]
```diff
@@ -21,21 +21,14 @@
...
-    @Test
-    public void testFunctionPointerReturningStruct() {
-        Path outputPath = getOutputFilePath("7903239_gen");
-        Path headerFile = getInputFilePath("test7903239.h");
-        run("--output", outputPath.toString(), headerFile.toString()).checkSuccess();
-        // nothing is generated that we can check, so we just check that jextract ran successfully
-        TestUtils.deleteDir(outputPath);
-    }
-}
+struct Foo foo = { a, b };
```
Comment (mcimadamore) [inline test/jtreg/generator/test7903239/libTest7903239.c:34]: Watch out: newline needed at the end

**Review by mcimadamore — APPROVED** [review]

**Comment by mcimadamore** [comment]
I have not sponsored this yet as the target branch of this PR seems wrong (e.g. `master` instead of `panama` or `jdk19`).

**Comment by lost-illusi0n** [comment]
Woops, should be good now.

**Pull request closed without merging**